### ISSUE TITLE
Update cookbook for locale routing parameter

### ISSUE THREAD

**Issue by xtreamwayz** [zendframework/zend-expressive#350]
I've followed the guidelines for setting up a locale depending on a [routing parameter](https://zendframework.github.io/zend-expressive/cookbook/setting-locale-depending-routing-parameter/).

The problem I have that if a locale is given in the url, I have to inject it inside every route. This has to be done after the LocalizationMiddleware. Since the config and routes are already loaded I found no way to set the current locale as the default locale for every route.

However by a simple hack this can be done automatically:

``` php
<?php

namespace App\Core\Router;

class FastRouteRouter extends \Zend\Expressive\Router\FastRouteRouter
{
    public function generateUri($name, array $substitutions = [])
    {
        if (!array_key_exists('locale', $substitutions)) {
            $substitutions['locale'] = substr(\Locale::getDefault(), 0, 2);
        }

        return parent::generateUri($name, $substitutions);
    }
}
```

Would this be useful to implement this directly or add it to the cookbook?

**Comment by moderndeveloperllc** [comment]
They seem to want all the locale stuff in the cookbook. If they ever [merge this pull request](https://github.com/zendframework/zend-expressive/pull/303/files), you can take out the `substr()` as the locale logic will account for language-country locales. (But not script - sorry Serbians!)

**Comment by xtreamwayz** [comment]
@Ocramius I'm setting it with middleware first:

```php
class LocalizationMiddleware
{
    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next = null)
    {
        $locale = $request->getAttribute(
            'locale',
            Locale::acceptFromHttp($request->getServerParams()['HTTP_ACCEPT_LANGUAGE'])
        );
        Locale::setDefault($locale);

        if ($request->getUri()->getPath() === '/') {
            return new RedirectResponse('/' . substr(\Locale::getDefault(), 0, 2));
        }

        return $next($request, $response);
    }
}
```

Thanx for the advice on the delegator factory. I'll try that.

**Comment by Ocramius** [comment]
> `Locale::setDefault($locale);`

Argh, we should really kill that stuff :-\

**Comment by xtreamwayz** [comment]
As discussed on IRC, it's better to inject the locale as request attribute and not rely on a static global.

**Comment by RalfEggert** [comment]
@xtreamwayz 

could you document that in the cookbook? Or is that already included in the recipies?

**Comment by xtreamwayz** [comment]
I'm figuring out a solution right now. The cookbook sets the static default locale and hasn't any locale detection.

What I have so far is something ugly but it's working: https://gist.github.com/xtreamwayz/1df70d2c072021633405a31e48f9138c
- It detects the locale if it's missing.
- Uses the request attribute to store it.
- Reuses the locale from the route result in the UrlHelper (the ugly hack). 

I still need to hook it up with a translation lib and twig. Once I have it completely working I'll update the cookbook.

**Comment by xtreamwayz** [comment]
I've got translations working with Twig and Symfony Translator. It should be almost the same for the Zend translator since they have almost the same api.

I've updated my gist: https://gist.github.com/xtreamwayz/1df70d2c072021633405a31e48f9138c

If you guys want to review it before I transfer it to the cookbook.

/cc @Ocramius @RalfEggert

**Comment by xtreamwayz** [comment]
This is merged with #351.
